### PR TITLE
Caching yarn and pnpm dependencies in Docker

### LINKED ISSUE #43167: Implement caching of NPM in CI  / local dev
Currently when NPM packages are installed on CI they are installed from the scratch by "compile www assets" pre-commit. 

This causes fairly frequent, intermittent issues when there is a networking issue or NPM server issue - for example:

```
Compile www assets (manual)..............................................Failed
  - hook id: compile-www-assets
  - exit code: 1
  
  yarn install v1.22.21
  (node:490) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
  (Use `node --trace-deprecation ...` to show where the warning was created)
  [1/4] Resolving packages...
  [2/4] Fetching packages...
  [] 0/1573[] 5/1573[] 11/1573[] 18/1573[] 26/1573[] 34/1573[] 41/1573[] 45/1573[] 53/1573[] 61/1573[] 68/1573[] 75/1573[] 84/1573[] 90/1573[] 97/1573[] 104/1573[] 111/1573[] 116/1573[] 122/1573[] 128/1573[] 135/1573[] 140/1573[] 147/1573[] 153/1573[] 160/1573[] 167/1573[] 176/1573[] 185/1573[] 192/1573[] 197/1573[] 198/1573[] 202/1573[] 206/1573[] 211/1573[] 216/1573[] 223/1573[] 229/1573[] 232/1573[] 234/1573[] 238/1573[] 241/1573[] 243/1573[] 246/1573[] 248/1573[] 250/1573[] 255/1573[] 260/1573[] 265/1573[] 269/1573[] 276/1573[] 284/1573[] 286/1573[] 287/1573[] 290/1573[] 292/1573[] 295/1573[] 298/1573[] 305/1573[] 311/1573[] 316/1573[] 325/1573[] 331/1573[] 336/1573[] 342/1573[] 346/1573[] 349/1573[] 353/1573[] 359/1573[] 364/1573[] 368/1573[] 372/1573[] 378/1573[] 385/1573[] 391/1573[] 396/1573[] 402/1573[] 409/1573[] 414/1573[] 420/1573[] 427/1573[] 432/1573[] 437/1573[] 441/1573[] 443/1573[] 444/1573[] 445/1573[] 448/1573[] 451/1573[] 454/1573[] 457/1573[] 462/1573[] 467/1573[] 470/1573[] 473/1573[] 477/1573[] 481/1573[] 485/1573[] 490/1573[] 496/1573[] 503/1573[] 512/1573[] 519/1573[] 524/1573[] 526/1573[] 529/1573[] 531/1573[] 535/1573[] 539/1573[] 545/1573[] 548/1573[] 550/1573[] 553/1573[] 560/1573[] 567/1573[] 574/1573[] 580/1573[] 591/1573[] 599/1573[] 607/1573[] 616/1573[] 620/1573[] 624/1573[] 627/1573[] 630/1573[] 634/1573[] 641/1573[] 650/1573[] 658/1573[] 668/1573[] 676/1573[] 684/1573[] 692/1573error Error: https://registry.yarnpkg.com/@babel/plugin-syntax-logical-assignment-operators/-/plugin-syntax-logical-assignment-operators-7.10.4.tgz: Request failed "500 Internal Server Error"
      at ResponseError.ExtendableBuiltin (/opt/airflow/files/home/.cache/pre-commit/repocv9ljx22/node_env-22.2.0/lib/node_modules/yarn/lib/cli.js:696:66)
      at new ResponseError (/opt/airflow/files/home/.cache/pre-commit/repocv9ljx22/node_env-22.2.0/lib/node_modules/yarn/lib/cli.js:802:124)
      at Request.<anonymous> (/opt/airflow/files/home/.cache/pre-commit/repocv9ljx22/node_env-22.2.0/lib/node_modules/yarn/lib/cli.js:66218:16)
      at Request.emit (node:events:520:28)
      at module.exports.Request.onRequestResponse (/opt/airflow/files/home/.cache/pre-commit/repocv9ljx22/node_env-22.2.0/lib/node_modules/yarn/lib/cli.js:141751:10)
      at ClientRequest.emit (node:events:520:28)
      at HTTPParser.parserOnIncomingClient (node:_http_client:700:27)
      at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)
      at TLSSocket.socketOnData (node:_http_client:542:22)
      at TLSSocket.emit (node:events:520:28)
  info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
  Traceback (most recent call last):
    File "/opt/airflow/./scripts/ci/pre_commit/compile_www_assets.py", line 71, in <module>
      subprocess.check_call(["yarn", "install", "--frozen-lockfile"], cwd=os.fspath(www_directory))
    File "/usr/local/lib/python3.9/subprocess.py", line 373, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['yarn', 'install', '--frozen-lockfile']' returned non-zero exit status 1.
  
  Traceback (most recent call last):
    File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
      return _run_code(code, main_globals, None,
    File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
      exec(code, run_globals)
    File "/opt/airflow/files/home/.local/share/hatch/env/virtual/apache-airflow-build/lib/python3.9/site-packages/hatchling/__main__.py", line 6, in <module>
      sys.exit(hatchling())
    File "/opt/airflow/files/home/.local/share/hatch/env/virtual/apache-airflow-build/lib/python3.9/site-packages/hatchling/cli/__init__.py", line 26, in hatchling
      command(**kwargs)
    File "/opt/airflow/files/home/.local/share/hatch/env/virtual/apache-airflow-build/lib/python3.9/site-packages/hatchling/cli/build/__init__.py", line 82, in build_impl
      for artifact in builder.build(
    File "/opt/airflow/files/home/.local/share/hatch/env/virtual/apache-airflow-build/lib/python3.9/site-packages/hatchling/builders/plugin/interface.py", line 155, in build
      artifact = version_api[version](directory, **build_data)
    File "/opt/airflow/hatch_build.py", line 613, in build_standard
      run(cmd, cwd=work_dir.as_posix(), check=True, shell=True)
    File "/usr/local/lib/python3.9/subprocess.py", line 528, in run
      raise CalledProcessError(retcode, process.args,
  subprocess.CalledProcessError: Command '['pre-commit run --hook-stage manual compile-www-assets --all-files']' returned non-zero exit status 1.
  Error building Airflow packages
Error preparing Airflow package
```

We should - similarly to Python packages, implement caching strategy that should speed up the installation of npm packages on a clean CI runner, as well as local development instance - where previously installed packages could be stored in the cache and reused. This should increase both - speed of CI jobs an stability of them (and speed of local installation of npm packages).

### PR DESCRIPTION
**Pull request by bugraoz93**
closes: #43167

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by bugraoz93** [comment]
Hey @potiuk, 

I have started including the yarn dependency installation very similar to dependency caching. Most probably, I am still missing a lot of parts. I wanted to align with you on the approach before going too deep and include every detail :)

I think using generic caches from GitHub actions have also a couple of actions for caching. Some of them specifically to yarn, npm etc... I don't think we would fully cascade that to the local environment though since it's limited to using it in GitHub CI. Maybe we can consider caching in CI and locally as separate tasks which using GitHub Actions could be easier to maintain in the future.

I couldn't properly run the `inline_scripts_in_docker.py` to include it in the Dockfile. That's why CI is failing. I can take care of that part later after we have agreed on the approach and implementation phases.

Thanks for your time!

**Comment by potiuk** [comment]
> I couldn't properly run the inline_scripts_in_docker.py to include it in the Dockfile. That's why CI is failing. I can take care of that part later after we have agreed on the approach and implementation phases.

This should happen automatically with pre-commit you do not have to run it manually - just `git add` your changes and commit (after running `pre-commit install` once) then it will happen automatically for you.

**Comment by potiuk** [comment]
> I think using generic caches from GitHub actions have also a couple of actions for caching. Some of them specifically to yarn, npm etc... I don't think we would fully cascade that to the local environment though since it's limited to using it in GitHub CI. Maybe we can consider caching in CI and locally as separate tasks which using GitHub Actions could be easier to maintain in the future.

Yeah. That's something that we could do for regular installing of Yarn/NPM dependencies as part of CI jobs, but that will not work well for installing Yarn/NPM deps during the Dockerfile build. And yes - also we have a few places where we install it for local development as you wrote. But I think it's better to use specific caching in each case:

* Dockerfile -> there the "tip" caching is really, really efficient, it works great because single build on the server saves all the rebuilds locally - and we just pull single container layer with installed dependencies - which is WAY faster than pulling half of the internet and 1000 node modules individually.

* pre-commit - it installs node automatically without us doing anything (when language: node is used). And there we also easily controle node version `default_language_version:  python: python3 node: 22.2.0`. There caching is achieved by caching the whole pre-commit cache. And the nice thing is that pre-commit will use single node environment for all the pre-commit hooks that have same specification. We have quite a few of them - some for old UI with yarn, some for new with pnpm and some for a number of node tools. And here I think it already works well with current caching in CI and locally without any change from our side.

I think both of them will be difficult to implement with any of the "standard" actions and are generally "better" - and same in CI and locally.

So yeah I think your current approach is right.

**Comment by bugraoz93** [comment]
> > I couldn't properly run the inline_scripts_in_docker.py to include it in the Dockfile. That's why CI is failing. I can take care of that part later after we have agreed on the approach and implementation phases.
> 
> This should happen automatically with pre-commit you do not have to run it manually - just `git add` your changes and commit (after running `pre-commit install` once) then it will happen automatically for you.

Yeah, I found the script itself from pre-commit and ran it multiple times as pre-commit and as a raw Python call. I think this is something in my local setup then. I can sort that out. Thanks!

**Comment by bugraoz93** [comment]
Yeah, it will be difficult with them. I have checked a couple of them, and they are a bit limited. Even though they seem like they are properly caching, it's still not flexible enough to fit into multiple use cases.
Awesome to hear the approach looks right! Hope I will ping you with a good solution in a couple of days :) 
Thanks for the detailed answer!

**Comment by potiuk** [comment]
> Yeah, I found the script itself from pre-commit and ran it multiple times as pre-commit and as a raw Python call. I think this is something in my local setup then. I can sort that out. Thanks!

One watchout here - in order for the inlining to work with **NEW** script - you should make sure to add a comment with the right script path just before the script that is "inlined" (and ending comment after) - this is the way how the script finds where to inline it.

**Comment by potiuk** [comment]
Looks good, but we also have to handle UI dependencies (with pnpm)

**Comment by bugraoz93** [comment]
While reviewing the logs to ensure everything was running smoothly, I noticed an issue: I had accidentally reversed the -+e option 🤦‍♂️.

However, I encountered the following error:
>#50 2.616 /scripts/docker/install_yarn_dependencies_from_branch_tip.sh: line 21: yarn: command not found

This suggests that yarn was either not installed or unavailable in the environment when the script was executed.

This made me realize that `npm` and `yarn` should have been installed at the point when the script is executed. I identified the best place to install them using apt and to cache them just before switching to the `airflow` user. Since `yarn` installs `node_modules` to a local directory, I copied the installed dependencies to a cache directory for future use. Once the cache was utilized in the appropriate location, it was deleted during the execution of `compile_www_assets.py` to ensure it was not included in the final package.

> Looks good, but we also have to handle UI dependencies (with pnpm)

Indeed, thanks for pointing that out! I wanted to make sure everything was functioning as expected before implementing a similar solution with `pnpm`. That’s the next step I’ll tackle. Many thanks for the quick review!

File: Dockerfile [hunk 1]
```diff
@@ -1394,6 +1394,88 @@ while echo "Running"; do
 done
 EOF
 
+
```
Comment (potiuk) [inline Dockerfile]: We do not really need to have it in PROD dockerfile. PROD image is generally  (in most cases) installed from installed packages, not from sources - and those packages already have assets compiled. Only for local development, npm step/airflow installation is done from local sources, but this is not something we should do frequently and it does not require caching.

Comment (bugraoz93) [reply Dockerfile]: Removed from PROD dockerfile. Thanks for the details!

**Comment by bugraoz93** [comment]
There are too many commits, rebased rather than merge :) Could you please check again when you have time?

**Comment by potiuk** [comment]
Sorry for missing that one. Rebased it. It looks really good.

File: scripts/ci/pre_commit/compile_ui_assets.py [hunk 2]
```diff
@@ -68,10 +69,14 @@ def get_directory_hash(directory: Path, skip_path_regexp: str | None = None) ->
...
     env = os.environ.copy()
     env["FORCE_COLOR"] = "true"
-    subprocess.check_call(
-        ["pnpm", "install", "--frozen-lockfile", "--config.confirmModulesPurge=false"],
-        cwd=os.fspath(ui_directory),
-    )
+    if os.getenv("AIRFLOW_PRE_CACHED_PNPM_PACKAGES", "false") == "true":
+        # Copy pnpm-cache to node_modules from pnpm-cache
+        shutil.copytree(pnpm_node_modules_cache_directory, node_modules_directory, dirs_exist_ok=True)
+    else:
```
Comment (potiuk) [inline scripts/ci/pre_commit/compile_ui_assets.py:75]: One small tthing here. I think we should ALWAYS run the install step below. This covers case where the .lock file changed in the meantime and we just need to update it based on pre-cached node_modules

File: scripts/ci/pre_commit/compile_www_assets.py [hunk 3]
```diff
@@ -68,7 +69,11 @@ def get_directory_hash(directory: Path, skip_path_regexp: str | None = None) ->
         shutil.rmtree(dist_directory, ignore_errors=True)
     env = os.environ.copy()
     env["FORCE_COLOR"] = "true"
-    subprocess.check_call(["yarn", "install", "--frozen-lockfile"], cwd=os.fspath(www_directory))
+    if os.getenv("AIRFLOW_PRE_CACHED_YARN_PACKAGES", "false") == "true":
```
Comment (potiuk) [inline scripts/ci/pre_commit/compile_www_assets.py:72]: Same here.

**Review by potiuk — COMMENTED** [review]
Just one small comment

**Comment by potiuk** [comment]
Hmm. Actually I looked closer (something was wrong here)

I think this is (and I completely forgot about ) - a bit different issue. What we are currently doing - we do not have at all node /yarn/pnpm in Breeze. We are using and expecting is that `pre-commit` will manage the node pnmp/yarn environment for us.

```
      - id: compile-ui-assets
        name: Compile ui assets (manual)
        language: node
        stages: ['manual']
        types_or: [javascript, ts, tsx]
        files: ^airflow/ui/
        entry: ./scripts/ci/pre_commit/compile_ui_assets.py
        pass_filenames: false
        additional_dependencies: ['pnpm@9.7.1']
      - id: compile-ui-assets-dev
        name: Compile ui assets in dev mode (manual)
        language: node
        stages: ['manual']
        types_or: [javascript, ts, tsx]
        files: ^airflow/ui/
        entry: ./scripts/ci/pre_commit/compile_ui_assets_dev.py
        pass_filenames: false
        additional_dependencies: ['pnpm@9.7.1']
      - id: compile-www-assets
        name: Compile www assets (manual)
        language: node
        stages: ['manual']
        'types_or': [javascript, ts, tsx]
        files: ^airflow/www/
        entry: ./scripts/ci/pre_commit/compile_www_assets.py
        pass_filenames: false
        additional_dependencies: ['yarn@1.22.21']
      - id: compile-www-assets-dev
        name: Compile www assets in dev mode (manual)
        language: node
        stages: ['manual']
        'types_or': [javascript, ts, tsx]
        files: ^airflow/www/
        entry: ./scripts/ci/pre_commit/compile_www_assets_dev.py
        pass_filenames: false
        additional_dependencies: ['yarn@1.22.21']
```

So we do not even have anything in the image.


I think the idea of doing similar caching in image (my idea) was really wrong. We could move node + all things to the image but it's not really needed. 

I am not really sure if we should do much now - the intermittent issues with node did not appear recently - so maybe just abandoning it for now is a better idea.

**Comment by bugraoz93** [comment]
Thanks for the review! 

That's exactly what's the case, so I installed the node /yarn/ppm in the image. I tested with Breeze, too, since Breeze builds the Dockerfile.ci in local development. Aa, this only covers the local development case, though which may only speed up and not solve the entire problem.

I agree, I haven't seen the problem with node for a while now. Let's abandon this for now. Also, managing these dependencies even in more places would be an additional burden. Even let's say we managed the dependencies in the image and removed them from pre-commit, it would increase the image size as well as bring a lot of vulnerabilities. Keeping this in the pre-commit environment still makes more sense. I missed this one. Awesome catch!

**Comment by bugraoz93** [comment]
Closing this PR. If the problem appears again multiple times and bothers us, we can follow up again and discuss possible solutions. :)

**Pull request closed without merging**